tools/jsii-compliance/report.ts

```diff
@@ -0,0 +1,206 @@
+const tablemark = require('tablemark');
+
+const SUCCESS = '✅';
+const FAILURE = '❌';
+const NA = 'N/A';
+
+/**
+ * Determine whether or not a test case should be excluded from a specific language.
+ * An exclusion happens either when the entire language binding is excluded, or a specific test case.
+ *
+ * @param test the test.
+ * @param suite the suite.
+ * @param language the language.
+ */
+function isExcluded(
+  testCase: schema.TestCase,
+  suite: schema.Suite,
+  language: string,
+) {
+  const testExcluded =
+    testCase.exclusions && testCase.exclusions[language] ? true : false;
+  const bindingExcluded =
+    suite.exclusions && suite.exclusions[language] ? true : false;
+  return testExcluded || bindingExcluded;
+}
+
+/**
+ * Determines the status of a specific test case with respect to a specific language.
+ *
+ * @param testCase the test case.
+ * @param language test language.
+ * @param reports the reports collected from all language bindings.
+ */
+function determineTestStatus(
+  testCase: schema.TestCase,
+  language: string,
+  reports: Record<string, schema.Report>,
+) {
+  const report = reports[language];
+  const testResult = report
+    ? report[normalizeTestName(testCase.name)]
+    : undefined;
+
+  if (!testResult) return FAILURE;
+  return testResult.status === 'success' ? SUCCESS : FAILURE;
+}
+
+/**
+ * Given a test name, normalize it so it can be compared across different language bindings.
+ *
+ * @param testName the test name.
+ */
+function normalizeTestName(testName: string): string {
+  return testName.toUpperCase();
+}
+
+/**
+ * Given a compliance report, normalize its test names so they are comparable to the
+ * tests defined in the suite.
+ *
+ * @param report report
+ */
+function normalizeReport(report: schema.Report): schema.Report {
+  const normalized: schema.Report = {};
+  Object.keys(report).forEach((k) => {
+    normalized[normalizeTestName(k)] = report[k];
+  });
+  return normalized;
+}
+
+/**
+ * Validates that a language specific compliance report doesn't violate the suite.
+ *
+ * Possible violations are:
+ *
+ *   - A test exist in the report that doesn't exist in the suite definition.
+ *
+ * @param report the report.
+ * @param language the language.
+ * @param suite the suite.
+ */
+function validateReport(
+  report: schema.Report,
+  language: string,
+  suite: schema.Suite,
+) {
+  const testsInReport = Object.keys(report);
+
+  const testsInSuite = suite.testCases.map((t: schema.TestCase) =>
+    normalizeTestName(t.name),
+  );
+
+  // make sure every test in the language report exist in the suite.
+  // this prevents us from adding tests only to a specific language.
+  for (const test of testsInReport) {
+    if (!testsInSuite.includes(test)) {
+      console.error(
+        `Test '${test}' from ${language} report does not exist in the compliance suite. Please add it to the suite definition.`,
+      );
+      process.exit(1);
```

Comment:
I kinda like accumulating as many errors as possible before bailing out... Because otherwise you discover about 1 problem, fix it, and only then discover the next problem...

Reply:
will do.